tests/unittests/core/fileio/fileutilstest.cpp

```diff
@@ -0,0 +1,313 @@
+/*
+ * LibrePCB - Professional EDA for everyone!
+ * Copyright (C) 2013 LibrePCB Developers, see AUTHORS.md for contributors.
+ * https://librepcb.org/
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*******************************************************************************
+ *  Includes
+ ******************************************************************************/
+
+#include <gtest/gtest.h>
+#include <librepcb/core/exceptions.h>
+#include <librepcb/core/fileio/filepath.h>
+#include <librepcb/core/fileio/fileutils.h>
+#if defined(Q_OS_WIN32) || defined(Q_OS_WIN64)  // Windows
+#include <fileapi.h>
+#endif
+
+#include <QtCore>
+
+/*******************************************************************************
+ *  Namespace
+ ******************************************************************************/
+namespace librepcb {
+namespace tests {
+
+/*******************************************************************************
+ *  Test Class
+ ******************************************************************************/
+
+class FileUtilsTest : public ::testing::Test {
+protected:
+  FilePath root;
+  FilePath rootFile;  // source for all operations
+  FilePath rootFileHidden;
+
+  FilePath subdir;
+  FilePath subdirFile;
+  FilePath subdirSubdir;
+  FilePath subdirSubdirFile;
+  FilePath subdirSubdirFileHidden;
+
+  FilePath rootFileMissing;
+  FilePath rootFileCopy;
+  FilePath subdirCopy;
+  FilePath subdirCopyFile;
+  FilePath subdirCopySubdir;
+  FilePath subdirCopySubdirFile;
+  FilePath subdirCopySubdirFileHidden;
+
+  QStringList filter;
+
+  static void SetupFile(const FilePath& pth, const QByteArray& content) {
+    QFile f(pth.toNative());
+    if (f.open(QIODevice::ReadWrite | QIODevice::NewOnly)) f.write(content);
+    f.close();
+  }
+
+  void SetUp() override {
```

Comment:
For consistency with all other tests, could you use the constructor/destructor instead of setup/teardown functions? It would also allow to directly initialize members like `filter` right in the constructor's initialization list.

Btw, I also wonder if it actually makes sense to create all these objects and files for each of the tests, although most tests only need one or two of them. Especially on Windows creating files is quite slow so it unnecessarily increases test time (although not too much). I think I'd have moved some of these things right into the corresponding test functions, but I'm also fine with keeping it that way.

Reply:
No, there is intent to make the test this way! 

I need for each test to have clean folder structure because the functions have *side effects*. I like tools like `gtest-parallel` which rapidly speeds up testing by starting all tests **in parallel** on machines that supports multiple cores (and this is the future of PCs) and there is no guide in the `gtest` in which order the tests should be run. I write always all tests compatible with this idiom (even when your code doesn't use it) because in the future it can be simply implemented without change in code of test as the test base will grow.

Other tests (like `FilePath`) have no side effects, so they don't need to use this idiom, they could use your way. Your lib is clever designed so the objects have little or no side effects and when I looked at some tests, they can be run in either order without problems (even in parallel)

According to `filter`, yes that can be made in initialization of variable, i'll move it. It's result of digging what the hell the function returns when I search for bug in your codebase.

Reply:
Hm, according [googletest FAQ](https://google.github.io/googletest/faq.html#CtorVsSetUp) the fixture object is explicitly created/destroyed for every single test, so the constructor/destructor behave the same way as SetUp/TearDown (with very small differences), or am I missing something?

Reply:
Yeah, from that point of view, implementing YAGNI, you are right. 

I made those tests using `INSTANTIATE_TEST_SUITE_P` macro, and those tests was only single parametrized test. In that scenario, the constructor is called only once per all parametrized items. (well the behavior on the page isn't right) Therefore I needed the SetUp/TearDown overrides.

Second assumtpion was, when SetUp/TearDown crashes (due to filesystem rejecting file creation / out of space - it is obvious problem on my system as you know, and that was my case), it is only failing test case with different message. The filesystem exception can't be handled in destructor or constructor (during cleaning or creation) and that terminates the test completely. 

This comes from my practice that file i/o *may* throw so I always suppose on first sight that they may throw, even in this scenario probably they may not. 

Implementing YAGNI, rewriting to `std::fstream` it can be done using constructors as You said, so I'll rewrite that way.